python/GafferSceneTest/VisibleSetTest.py

```diff
@@ -0,0 +1,202 @@
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import IECore
+
+import GafferSceneTest
+import GafferScene
+
+class VisibleSetTest( GafferSceneTest.SceneTestCase ) :
+
+	def testExpansions( self ) :
+
+		e = GafferScene.VisibleSet()
+		e.expansions = IECore.PathMatcher( [ "/a", "/b", "/a/b", "/b/c", "/a/b/c", "/c/d/e", "/c/d", "/d" ] )
+
+		self.assertEqual( e.match( "/a" ), ( IECore.PathMatcher.Result.ExactMatch | IECore.PathMatcher.Result.DescendantMatch ) )
+		self.assertEqual( e.match( "/b" ), ( IECore.PathMatcher.Result.ExactMatch | IECore.PathMatcher.Result.DescendantMatch ) )
+		self.assertEqual( e.match( "/d" ), IECore.PathMatcher.Result.ExactMatch )
+		self.assertEqual( e.match( "/a/b" ), IECore.PathMatcher.Result.EveryMatch )
+		self.assertEqual( e.match( "/b/c" ), ( IECore.PathMatcher.Result.ExactMatch | IECore.PathMatcher.Result.AncestorMatch ) )
+		self.assertEqual( e.match( "/a/b/c" ), ( IECore.PathMatcher.Result.ExactMatch | IECore.PathMatcher.Result.AncestorMatch ) )
+
+		# Although "/c/d" is a member of `expansions` we return NoMatch as its parent "/c" is not expanded.
+		self.assertEqual( e.match( "/c/d" ), IECore.PathMatcher.Result.NoMatch )
+
+		## \todo "/c/d/e" returns an ExactMatch and AncestorMatch as it and its parent "/c/d" are members of `expansions`, but in
+		# practice neither of these locations would be visible as "/c" is not in `expansions`.
+		# Updating IECore.PathMatcher to be able to test whether all ancestors match could give us an efficient way of
+		# returning NoMatch in this situation.
+		self.assertEqual( e.match( "/c/d/e" ), ( IECore.PathMatcher.Result.ExactMatch | IECore.PathMatcher.Result.AncestorMatch ) )
+
+		# While "/c" has descendants in `expansions`, we do not return a DescendantMatch as "/c" itself is not in `expansions` so
+		# its descendants cannot be visible.
+		self.assertEqual( e.match( "/c" ), IECore.PathMatcher.Result.NoMatch )
```

Comment:
Agreed, though we're stuck in a bit of an awkward transition where the RenderController would introduce that unwanted extra level of expansion. I've left todos in 46e4e68 to sort this out once the RenderController is updated to ignore siblings of visible things and we can treat these results as visibility as nature intended, rather than expansion.